android/sdl_android/src/androidTest/java/com/smartdevicelink/managers/screen/choiceset/PresentChoiceSetOperationTests.java

```diff
@@ -102,4 +119,114 @@ private KeyboardProperties getKeyBoardProperties(){
+
+				RPCResponse response = new RPCResponse(FunctionID.CANCEL_INTERACTION.toString());
+				response.setSuccess(true);
+				cancelInteraction.getOnRPCResponseListener().onResponse(0, response);
+
+				return null;
+			}
+		};
+		doAnswer(cancelInteractionAnswer).when(internalInterface).sendRPC(any(CancelInteraction.class));
+
+		choiceSet.cancel();
+
+		assertTrue(presentChoiceSetOperation.isExecuting());
+		assertFalse(presentChoiceSetOperation.isFinished());
+	}
+
+	public void testCancelingChoiceSetUnsuccessfullyIfThreadIsRunning(){
+		presentChoiceSetOperation.run();
+
+		Answer<Void> cancelInteractionAnswer = new Answer<Void>() {
+			@Override
+			public Void answer(InvocationOnMock invocation) {
+				Object[] args = invocation.getArguments();
+				CancelInteraction cancelInteraction = (CancelInteraction) args[0];
+
+				assertEquals(cancelInteraction.getCancelID(), Test.GENERAL_INTEGER);
+				assertEquals(cancelInteraction.getInteractionFunctionID().intValue(), FunctionID.PERFORM_INTERACTION.getId());
+
+				RPCResponse response = new RPCResponse(FunctionID.CANCEL_INTERACTION.toString());
+				response.setSuccess(false);
+				cancelInteraction.getOnRPCResponseListener().onResponse(0, response);
+
+				return null;
+			}
+		};
+		doAnswer(cancelInteractionAnswer).when(internalInterface).sendRPC(any(CancelInteraction.class));
+
+		choiceSet.cancel();
+
+		assertTrue(presentChoiceSetOperation.isExecuting());
+		assertFalse(presentChoiceSetOperation.isFinished());
+	}
+
+	public void testCancelingChoiceSetIfThreadHasFinished(){
+		presentChoiceSetOperation.run();
+		presentChoiceSetOperation.finishOperation();
+
+		choiceSet.cancel();
+
+		verify(internalInterface, never()).sendRPC(any(CancelInteraction.class));
+
+		assertFalse(presentChoiceSetOperation.isExecuting());
+		assertTrue(presentChoiceSetOperation.isFinished());
+	}
+
+	public void testCancelingChoiceSetIfThreadHasNotYetRun(){
+		choiceSet.cancel();
+
+		verify(internalInterface, never()).sendRPC(any(CancelInteraction.class));
+
+		// Once the thread has started
+		presentChoiceSetOperation.run();
+
+		// Make sure neither a `CancelInteraction` or `PerformInteraction` RPC is sent
+		verify(internalInterface, never()).sendRPC(any(CancelInteraction.class));
+		verify(internalInterface, never()).sendRPC(any(PerformInteraction.class));
+
+		assertFalse(presentChoiceSetOperation.isExecuting());
+		assertTrue(presentChoiceSetOperation.isFinished());
+	}
+
+	public void testCancelingChoiceSetIfHeadUnitDoesNotSupportFeature(){
+		// Only supported with RPC spec versions 6.0.0+
+		presentChoiceSetOperation.sdlMsgVersion = new SdlMsgVersion(5, 3);
+		presentChoiceSetOperation.run();
+
+		choiceSet.cancel();
+
+		verify(internalInterface, never()).sendRPC(any(CancelInteraction.class));
+	}
+
+	public void testCancelingChoiceSetIfHeadUnitDoesNotSupportFeatureButThreadIsNotRunning(){
```

Comment:
I see an issue with interrupting threads to cancel operations. Basically, we have three different threads `SdlService` thread, `Executor` thread, `SdlProxyBase` dispatcher thread. They can all run code in the operation class at some point which can cause some issues.

For example, the `SdlService` thread can call `choiceSet.cancel()` which interrupts that `SdlService` thread and not the `Executor` thread. However, when we check if the thread is interrupted in `AsynchronousOperation.isCancelled()`, it checks if the current thread is interrupted and because the `Executor` is running the operation.run(), it checks if that `Executor` thread is interrupted and it will be not be interrupted because it is a different thread.
 
That issue can clearly be seen if you run the operation in this test `testCancelingChoiceSetIfHeadUnitDoesNotSupportFeatureButThreadIsNotRunning()` in another thread (which what happens in an actual app)

try to replace:
```
presentChoiceSetOperation.run();
```
with this
```
new Thread(presentChoiceSetOperation).start();
```

You will see the test fails. I suggest not using thread interruptions cause they may lead to unpredictable scenarios.